producer/buffer/buffer.go

```diff
@@ -0,0 +1,215 @@
+	}
+}
+
+type buffer struct {
+	sync.RWMutex
+
+	buffers       *list.List
+	opts          Options
+	maxBufferSize uint64
+	onFinalizeFn  data.OnFinalizeFn
+	m             bufferMetrics
+
+	size     *atomic.Uint64
+	doneCh   chan struct{}
+	isClosed bool
+}
+
+// NewBuffer returns a new buffer.
+func NewBuffer(opts Options) producer.Buffer {
+	if opts == nil {
+		opts = NewBufferOptions()
+	}
+	b := &buffer{
+		buffers:       list.New(),
+		maxBufferSize: uint64(opts.MaxBufferSize()),
+		opts:          opts,
+		m:             newBufferMetrics(opts.InstrumentOptions().MetricsScope()),
+		size:          atomic.NewUint64(0),
+		doneCh:        make(chan struct{}),
+		isClosed:      false,
+	}
+	b.onFinalizeFn = b.subSize
+	return b
+}
+
+func (b *buffer) Buffer(d producer.Data) (producer.RefCountedData, error) {
+	b.Lock()
+	if b.isClosed {
+		b.Unlock()
+		return nil, errBufferClosed
+	}
+	var (
+		dataSize      = uint64(d.Size())
+		maxBufferSize = b.maxBufferSize
+	)
+	if dataSize > maxBufferSize {
+		b.Unlock()
+		b.m.dataTooLarge.Inc(1)
+		return nil, errInvalidDataLargerThanMaxBufferSize
+	}
+	targetBufferSize := maxBufferSize - dataSize
+	if b.size.Load() > targetBufferSize {
+		if err := b.produceOnFullWithLock(targetBufferSize); err != nil {
+			b.Unlock()
+			return nil, err
+		}
+	}
+	b.size.Add(dataSize)
+	rd := data.NewRefCountedData(d, b.onFinalizeFn)
+	b.buffers.PushBack(rd)
+	b.Unlock()
+	return rd, nil
+}
+
+func (b *buffer) produceOnFullWithLock(targetSize uint64) error {
+	switch b.opts.OnFullStrategy() {
+	case ReturnError:
+		return errBufferFull
+	case DropEarliest:
+		b.dropEarliestUntilTargetWithLock(targetSize)
+	}
+	return nil
+}
+
+func (b *buffer) dropEarliestUntilTargetWithLock(targetSize uint64) {
+	var next *list.Element
+	for e := b.buffers.Front(); e != nil && b.size.Load() > targetSize; e = next {
+		next = e.Next()
+		d := e.Value.(producer.RefCountedData)
+		if !d.IsDroppedOrConsumed() {
+			b.m.messageDropped.Inc(1)
+			b.m.bytesDropped.Inc(int64(d.Size()))
+			d.Drop()
+		}
+		b.buffers.Remove(e)
+	}
+}
+
+func (b *buffer) Init() {
+	go b.cleanupForever()
+}
+
+func (b *buffer) cleanupForever() {
+	ticker := time.NewTicker(b.opts.CleanupInterval())
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			b.Lock()
+			b.cleanupWithLock()
+			b.m.messageBuffered.Update(float64(b.buffers.Len()))
+			b.Unlock()
+			b.m.bytesBuffered.Update(float64(b.size.Load()))
+		case <-b.doneCh:
+			return
+		}
+	}
+}
+
+func (b *buffer) cleanupWithLock() {
+	var next *list.Element
+	for e := b.buffers.Front(); e != nil; e = next {
+		next = e.Next()
+		d := e.Value.(producer.RefCountedData)
+		if d.IsDroppedOrConsumed() {
+			b.buffers.Remove(e)
+		}
+	}
+}
+
+func (b *buffer) Close() {
+	// Stop taking writes right away.
+	b.Lock()
+	if b.isClosed {
+		b.Unlock()
+		return
+	}
+	b.isClosed = true
+	b.Unlock()
+
+	b.waitUntilAllDataConsumed()
+	close(b.doneCh)
+}
+
+func (b *buffer) waitUntilAllDataConsumed() {
```

Comment:
Can you get into a situation where this can run forever? e.g., the consumer is not consuming as fast as the data is produced? If so, does that mean it can hold up the shutdown sequence of a process?

Reply:
This only gets called after Close is called, which will stop taking in new writes and wait for buffered writes to be consumed, but there could still be case where down stream consumer is not consuming data, in which case the user of the producer library could put a timeout on the `Close()` just like what we did in all the graceful shutdowns.

Reply:
Could that potentially cause metric loss if we forcefully kill the process before all data are consumed?

Reply:
Yeah if you don't wait until the return of the Close that is metric loss, otherwise you have to Close to stop taking in new writes and keep waiting.